pkg/kv/kvserver/spanset/batch.go

```diff
@@ -799,83 +814,151 @@ func (s spanSetBatch) NewBatchOnlyMVCCIterator(
 	return NewIteratorAt(mvccIter, s.spans, s.ts), nil
 }
 
-func (s spanSetBatch) Commit(sync bool) error {
-	return s.b.Commit(sync)
+// shallowCopy returns a shallow copy of the spanSetBatch. The returned batch
+// shares the same underlying storage.Batch but has its own spanSetBatch wrapper
+// with a new copy of the SpanSet that uses a shallow copy of the underlying
+// spans.
+func (s spanSetBatch) shallowCopy() *spanSetBatch {
+	b := s
+	b.spanSetReader.spans = b.spanSetReader.spans.ShallowCopy()
+	b.spanSetWriteBatch.spanSetWriter.spans = b.spanSetWriteBatch.spanSetWriter.spans.ShallowCopy()
+	b.spans = b.spans.ShallowCopy()
+	return &b
+}
+
+// spanSetWriteBatch wraps a storage.WriteBatch and adds span checking.
+type spanSetWriteBatch struct {
```

Comment:
nit: put this type above `spanSetBatch`, as this file is ordered from more basic to bigger interfaces. This might have also (or not) made the diff cleaner / easier to review (mostly signature changes).